src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -170,6 +173,35 @@ internal Cci.IFieldReference GetModuleVersionId(Cci.ITypeReference mvidType)
             return _mvidField;
         }
 
+        internal Cci.IFieldReference GetInstrumentationPayload(int analysisKind, Cci.ITypeReference payloadType)
+        {
+            InstrumentationPayloadField payloadField;
+            if (!_instrumentationPayloadFields.TryGetValue(analysisKind, out payloadField))
+            {
+                payloadField = new InstrumentationPayloadField(this, analysisKind, payloadType);
+                if (!_instrumentationPayloadFields.TryAdd(analysisKind, payloadField))
+                {
+                    // Another thread added the field.
+                    payloadField = _instrumentationPayloadFields[analysisKind];
+                }
+            }
+
+            return payloadField;
+        }
+
+        internal ImmutableArray<InstrumentationPayloadField> GetInstrumentationPayloads()
+        {
+            Debug.Assert(IsFrozen);
+            int payloadsLength = _instrumentationPayloadFields.Keys.Max() + 1;
```

Comment:
It feels like an overkill to allocate an ImmutableArray for this purpose, especially with potentially unused elements, which consumers have to account for. Concurrent dictionary implements IReadOnlyCollection interface, perhaps this method should simply return it. The consumer will be able to know the Count and to enumerate pairs.